### ISSUE TITLE
Can We Do More To Preserve Combined Errors?

### ISSUE THREAD

**Issue by adamgfraser** [zio/zio#2693]
Right now if we have a `Cause` composed of multiple `Fail` cases, we risk losing information when we use most error handling combinators. For example:

```scala
ZIO.halt(Cause.Then(Cause.fail("foo"), Cause.fail("bar"))).catchAll(ZIO.fail)
// Returns ZIO.fail("foo")
```

`foldM`, which almost all error handling combinators are derived from, uses `Cause.failureOrCause` internally which returns the first checked exception, if one exists.

I'm not sure what to do here as in a way it is an inherent limitation of allowing people to interact with the error channel as if a single `E` exists when in fact multiple `E` values can exist, but it seems less than ideal and is something we will have to think about more as we try to use composed causes, so wanted to raise the issue to see if others have thoughts.

**Comment by iravid** [comment]
I'm in agreement with the general sentiment of the issue. However, I don't think we should automatically aggregate failures for users in general combinators like `catchAll`. My intuition says that this sort of feature is more suitable for enhancing the execution trace, whereas error aggregation is something the user should choose to do explicitly (and we should provide convenient combinators for doing so).

**Comment by adamgfraser** [comment]
I tend to agree with that. My concern is more about the risk of surprise when we do create combined errors and the implication for our ability to actually return combined errors from our own combinators. For example in #2689 I made `tapError` return a `Cause.then(e1, e2)` when both the original effect and the tap effect fail. That seemed like a logical, value preserving thing to do. But then if a user calls that method and recovers with something like `catchAll` they may not even realize an error occurred in the tap effect.

**Comment by regiskuckaertz** [comment]
I don’t think effects in combinators like `tapError` should be allowed to fail. Perhaps the use of `fold` should be discouraged 🤔 There are many PRs where it ended up yielding bugs in corner cases and has to be replaced with `foldCause`

**Comment by adamgfraser** [comment]
@regiskuckaertz I tend to think you are right. `tapError` is very similar to combinators like `ensuring` in that it creates the possibility that both the original effect and the tap / ensuring effect would fill fail. We've dealt with that in combinators like `ensuring` by requiring them not to fail and I think we should probably do the same here since even though we can preserve both errors with `Cause` it is very easy to then lose the second error. Unless anyone has other thoughts I will submit a PR to change that.

**Comment by adamgfraser** [comment]
In `tap` we are guaranteed that if the tap effect begins executing the original effect has completed successfully, so there can be at most one error (the error from the tap effect).

In contrast with `tapError` there is already one original error and then without the proposed restriction the tap effect could fail with its own error so we would have two errors, forcing us to use something like `Cause.Then` to encode both errors and creating the risk of the second error being lost with standard error recovery combinators.

**Comment by iravid** [comment]
The same could be applied to catchAll.

Put differently, I am not convinced this is equivalent to finalizers.

**Comment by adamgfraser** [comment]
I think there is actually a significant difference with `catchAll`.

The contract of `catchAll` is essentially "transform the error value with the specified function" so `catchAll` is always going to throw away the old error and return a new effect based on that. It unambiguously is only going to return the result of the `catchAll` effect, whether it is a success or failure.

In contrast, the contract of `tapError` is I think "do something on the side with the error and let me proceed with the computation". So it is different in creating the possibility of two different errors that should both be retained.

**Comment by iravid** [comment]
Uhh, you could equivalently argue that the contract should be that if the catchAll handler fails, both errors should be preserved

**Comment by iravid** [comment]
Are there any other combinators, apart from tapError, that exhibit the same ambiguity?

**Comment by adamgfraser** [comment]
So sorry, I think we agree that in `catchAll` if the handler succeeds the overall effect should succeed with the result of the handler and if the handler fails the effect should fail with the result of the handler.

For `tapError`, if the tap succeeds the overall effect should fail with the error of the original effect. If the tap fails what should we fail with?

**Comment by adamgfraser** [comment]
I think `tapCause` is the only one that exhibits the same ambiguity. The pattern is where do something that is "on the side" of the normal data flow and is not a finalizer, which I think is just tap variants.

**Comment by iravid** [comment]
Im of the opinion that for tapError and tapCause failing with the new cause is a perfectly fine choice.

Aggregating errors with `Then` is something we need to investigate more thoroughly in terms of the affect on user ergonomics; we have only ever used it (if I’m not mistaken) for preserving finalizer defects.

If this is only about tapError/Cause, I wouldn’t mind making them infallible, but I do feel that’s a bit inconsistent.

**Comment by iravid** [comment]
A while ago I thought that Cause.Then#failureOrOption should give you the latest failure rather than the earliest, but I think that breaks other things

**Comment by adamgfraser** [comment]
I tend to agree with you that using `Cause.Then` here was a mistake on my part and should be reverted.

Saying we will just fail with the tap error is perfectly logical. I think sometimes it could result in counterintuitive results. For example, `ZIO.fail("Invalid user").tapCause(_ => ZIO.fail("Logging service down"` would return an effect that failed with the logging error but not the original user error. But to your point you could argue that the user has the ability to incorporate the old error into the new error if they want to.

Where it becomes a little less clear is in `tapCause`, because now you have an original cause that could be an interruption or a die failure. So if you follow the original logic above if the tap effect fails with a checked exception you could inadvertently convert an unchecked exception to a checked exception.
